Atualização do README com novas seções: Resultados e Discussões e Conclusão (#4)
SteffanoP/rl-expected-sarsa-offpolicy: README.md

- **Estados**: Posição (x,y) e velocidade (vx,vy).
- **Ações**: 9 possíveis acelerações (-1,0,1 em cada eixo).
- **Recompensas**: -1 por passo, -5 por colisão, 0 ao cruzar linha de chegada.

## Resultados e Discussões

Esta seção discute os resultados obtidos a partir dos testes realizados em cada um dos ambientes listados previamente, a fim de exibir o impacto que as três políticas utilizadas inferem ao serem aplicadas. Além disso, buscamos evidenciar se é possível melhorar o desempenho do Expected-SARSA ao usar outras políticas de treinamento (_behavior policies_).

Para podermos ter uma visão apropriada dos cenários e como que as políticas estão atuando neles, a otimização dos nossos hiperparâmetros com o Optuna foi essecial para tal. O Expected-SARSA chegou a variar consideravelmente durante os nossos testes, o que culminava em resultados divergentes. Entretanto, quando conseguimos encontrar um intervalo de valores padrão com o Optuna, os resultados passaram a ser mais condizentes e satisfatórios ao utilizarmos por padrão o `max_passos` configurado em `9000`.

> Na Imagem 1 apresentamos o resultado comparativo obtido ao executar as três políticas no ambiente do RaceTrack. Fomos observamos ao longo dos treinos que a política epsilon-gulosa apresenta um melhor desempenho no ambiente exceto quando executamos o grupo "3) TODOS x softmax", pois a softmax ultrapassa as de mais. Como comparativo final, obtemos uma divisão nos resultados: a _epsilon-gulosa_ possui algumas vantagens, porém ao visualizar o último gráfico, percebemos que a _softmax_ tem melhor adaptação ao cenário.

Reviewer: ```suggestion
Na Imagem 1 apresentamos o resultado comparativo obtido ao executar as três políticas no ambiente do RaceTrack. Observamos ao longo dos treinos que a política epsilon-gulosa apresenta um melhor desempenho no ambiente exceto quando executamos o grupo "3) TODOS x softmax", pois a softmax ultrapassa as demais. Como comparativo final, obtemos uma divisão nos resultados: a _epsilon-gulosa_ possui algumas vantagens, porém ao visualizar o último gráfico, percebemos que a _softmax_ tem melhor adaptação ao cenário.
```

Ao executarmos o comparativo no ambiente do Cliffwalking, temos a Imagem 2 que demonstra um contraponto ao ambiente anterior, pois a política _epsilon-gulosa_ se destaca perante o desempenho das outras aplicadas. O aprendizado da política em questão oferece melhor adaptabilidade ao cenário nos três grupos treinados quando apresenta menos chances do agente cair no precipício. Aqui, ele aprende rapidamente a melhor política a usar, pois a forma "gulosa" acaba proporcionando isso devido ao Cliffwalking ser mais simples do que o ambiente anteriormente mencionado.

![a imagem mostra o gráfico do ambiente Cliffwalking, onde é possível uma comparação do uso das políticas nesse ambiente](/assets/Cliffwalking.png)Imagem 2. Comparativo das políticas no ambiente do Cliffwalking

Os resultados obtidos na Imagem 3, referem-se à execução das políticas no ambiente do FrozenLake. Assim, como no Cliffwalking, a política com melhor desempenho no FrozenLake foi a _epsilon-gulosa_. Aqui podemos inferir uma situação semelhante à anterior, pois o ambiente atual é ainda mais simples que o Cliffwalking, haja vista que seu resultado é binário variando entre 0 (caiu no buraco) e 1 (alcançou o presente). 

> Um detalhe curioso obtido nos treinamentos iniciais desse ambiente foi a exibição difusa nos gráficos. Como o resultado do FrozenLake é binário, os gráficos apresentavam apenas linhas retas verticais que variavam de 0 à 1. Nesse contexto, tivemos que utilizar a função `smooth` do Python para ajudarnos a ter uma melhor e mais adequada visualização dos dados.

Reviewer: ```suggestion
> [!NOTE]
> Um detalhe curioso obtido nos treinamentos iniciais desse ambiente foi a exibição difusa nos gráficos. Como o resultado do FrozenLake é binário, os gráficos apresentavam apenas linhas retas verticais que variavam de 0 à 1. Nesse contexto, tivemos que utilizar a função `smooth` do Python para ajudarnos a ter uma melhor e mais adequada visualização dos dados.
```

Reviewer: > [!NOTE]
> Isso é um exemplo de nota técnica, ou seja, de como vai ficar.

![a imagem mostra o gráfico do ambiente FrozenLake, onde é possível uma comparação do uso das políticas nesse ambiente](/assets/FrozenLake.png)Imagem 3. Comparativo das políticas no ambiente do FrozenLake

> Dado a exploração do ambiente Taxi, a Imagem 4 demonstra que o percurso com base nos 9000 passos executados retorna o melhor valor médio a partir da política _epsilon-gulosa_. Este ambiente não apresenta uma simplicidde como os dois anteriores, haja vista que possui mais ações e outra distribuição das recompensas, porém, verificamos que a forma "gulosa" trouxe a adaptação do melhor trajeto mais rapidamente.

Reviewer: ```suggestion
Dado a exploração do ambiente Taxi, a Imagem 4 demonstra que o percurso com base nos 9000 passos executados retorna o melhor valor médio a partir da política _epsilon-gulosa_. Este ambiente não apresenta uma simplicidade como os dois anteriores, haja vista que possui mais ações e outra distribuição das recompensas, porém, verificamos que a forma "gulosa" trouxe a adaptação do melhor trajeto mais rapidamente.
```

![a imagem mostra o gráfico do ambiente Taxi, onde é possível uma comparação do uso das políticas nesse ambiente](/assets/Taxi.png)Imagem 4. Comparativo das políticas no ambiente do Taxi

> A partir desse comparativo, observamos que a influencia e impacto da política pode variar conforme o ambiente e suas características, de uma forma mais evidente ou não. Entretanto, é vá-lido apontar que o resultado médico de cada um desses cenários é o que pode determinar a adaptabilidade do agente nas ações determinadas.

Reviewer: ```suggestion
A partir desse comparativo, observamos que a influência e impacto da política pode variar conforme o ambiente e suas características, de uma forma mais evidente ou não. Entretanto, é válido apontar que o resultado de cada um desses cenários é o que pode determinar a adaptabilidade do agente nas ações determinadas.
```

## Conclusão

> O estudo em questão foi desenvolvido a fim de demosntrar as adaptações que as políticas _epsilon-gulosa_, _softmax_ e _aleatoria_ podem desempenhar nos cenários de treinamento. Nesse contexto, destacamos que a otimização dos hiperparâmetros usados no código é crucial para obter resultados válidos com a finalidade de compará-los.

Reviewer: ```suggestion
O estudo em questão foi desenvolvido a fim de demosntrar as adaptações que as políticas _epsilon-gulosa_, _softmax_ e _aleatoria_ podem desempenhar nos cenários de treinamento. Nesse contexto, destacamos que a otimização dos hiperparâmetros, com o optuna, é crucial para obter resultados válidos com a finalidade de compará-los.
```
